### ISSUE TITLE
Markdown Formatter

### ISSUE THREAD

**Issue by MyGodItsFull0fStars** [Zettlr/Zettlr#3871]
### Checklist

- [X] Zettlr does not possess this feature (if it does, but the feature doesn't work out for you, propose an enhancement instead.)
- [X] Zettlr's current features cannot be used to approximate this behaviour (otherwise, propose an enhancement instead.)
- [X] I have searched the documentation to make sure I did not overlook it.
- [X] I have searched both open and closed issues to see if someone has already proposed this.

### Feature Proposal

Having an automatic Markdown formatter for Zettlr would enhance readability and consistency of notes.
These would be beneficial for users that already rely on such functionality in different editors and want to switch to Zettlr and miss this convenience (such as myself).
Also a major benefit would be that if a workspace is shared with multiple people, that a consistent style is enforced inside a project. 

### Example Workflow

Note: the ! before a # are used to cancel the styling of the preview.

### Create a newline between headers and content

This would consistently add a newline after a Header if there isn't one. Also could be used to enforce only using one newline between header and content, so if multiple newlines are in between, those could be reduced to use only one.

!# Header Name
Content Content Content...

Would turn to:

!# Header Name

Content Content Content...

### Enforce Newline Before Listing

The following does not properly display a list if there does not exist a new line between the content and the list items.

Content content...
* list item
* list item

Would turn to:

Content content...

* list item
* list item

### Caveats

It could happen that enforcing some formatting rules does break some existing notebooks or that display errors are encountered.

Also enforcing one "right" formatter could make some people unhappy if they have a personal preference, so maybe be able to customize the formatter would be beneficial.

### Do you wish to attempt implementing this yourself?

No

### Platforms

- [X] Windows
- [X] macOS
- [X] Linux

### Additional Information

I believe that there should either be an option to format on save, or as a shortcut.
Also that it should be added as a flag that can be turned on/off.

I really would like to develop this feature myself, but I do not have much proficiency with TypeScript and not much time on my hand at the moment. So if it is not implemented in the near future, I could take up the task of implementing a solution for this feature request.

**Comment by nathanlesage** [comment]
This sounds similar to my idea of adding a linter to the editor. Feel free to have a look around, maybe the reMark suite already offers a formatter? I don't know if it'll manage into 3.0 but it's an interesting idea.

**Comment by MyGodItsFull0fStars** [comment]
Thank you for your answer. 
Maybe the formatting could be added into the linter feature?
Yes, once I got more time on my hands, I will look around and see if I can help with that feature.

Were you referring to this remark project: https://github.com/remarkjs/remark?

**Comment by nathanlesage** [comment]
> Were you referring to this remark project: [remarkjs/remark](https://github.com/remarkjs/remark)?

Yup! I know they're doing a ton of work in the realm of Markdown, and I wanted to begin utilizing their work for parsing Markdown files more easily, and such. And yeah, whether or not we can integrate linter and formatter depends a little bit on the library, but in general we want both to be separate so that we can allow users to format specific parts based on the linter, or the full document (i.e. something like "format on save")

**Comment by MyGodItsFull0fStars** [comment]
Okay it makes sense that they both are separate. I was thinking more of that the formatting logic that manipulates the file could be used for both enforcing the linter rules and the Markdown formatter rules. Of course, if that is feasible to do so.

If that is the project, I will read through their functionality in my spare time, and maybe I can give it a try :)

**Comment by nathanlesage** [comment]
That would be awesome! :) I think the easiest right now would simply be a linter!
